modules/stochastic_tools/include/surrogates/SurrogateModel.h

```diff
@@ -0,0 +1,103 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "StochasticToolsApp.h"
+#include "GeneralUserObject.h"
+#include "SamplerInterface.h"
+#include "LoadSurrogateDataAction.h"
+
+class SurrogateModel : public GeneralUserObject, public SamplerInterface
+{
+public:
+  static InputParameters validParams();
+
+  SurrogateModel(const InputParameters & parameters);
+
+  /**
+   * Evaluate surrogate model given a row of parameters.
+   */
+  virtual Real evaluate(const std::vector<Real> & x) const = 0;
+
+  /**
+   * Train the surrogate model.
+   */
+  virtual void train() = 0;
+
+  ///@{
+  /**
+   * Initialize/Finalize that only execute when training.
+   */
+  virtual void trainInitialize() {}
+  virtual void trainFinalize() {}
+  ///@}
+
+  /**
+   * Return True if the object is performing training (i.e., filename is not set)
+   */
+  bool isTraining() const;
+
+  ///@{
+  /**
+   * Declare model data for loading from file as well as restart
+   */
+  template <typename T>
+  T & declareModelData(const std::string & data_name);
+
+  template <typename T>
+  T & declareModelData(const std::string & data_name, const T & value);
+  ///@}
+
+  /**
+   * UserObject methods that are used to implement the training/evaluate interfaces.
+   */
+  virtual void initialize() final;
+  virtual void execute() final;
+  virtual void finalize() final;
+  virtual void threadJoin(const UserObject &) final {} // GeneralUserObjects are not threaded
```

Comment:
It should be and I put it there years ago. But, someone came along and created ThreadedGeneralUserObject, which I think inherits from this, so had to take it out.

I thought about tackling that, but then decided I don't care anymore since I already did it once.

Reply:
How difficult would it be to have threading for our surrogate models? The extra level of parallelization could be beneficial.

Reply:
It is quite difficult and probably not worth the effort. The main items that are threaded in MOOSE are the element and node loops, which are well defined. What type of work and data and how it is used is not well defined for surrogates, so generalizing that case would be difficult. Furthermore, threading code is way more difficult than MPI code.